3.12/administration-reduce-memory-footprint.md

```diff
@@ -15,123 +15,445 @@ Do not apply any of the changes suggested here before you have tested them in
+operations if the cache is full and we have observed that RocksDB
+instances can get stuck in this case. You have been warned.
 
-There are additional options you can enable to avoid that the index and filter
-blocks get evicted from cache.
+There are a number of things for which RocksDB needs RAM. It is possible
+to make it so that all of these RAM usages count towards the block cache
+usage. This is usually sensible, since it effectively allows to keep
+RocksDB RAM usage under a certain configured limit (namely the block
+cache size limit). If these things are not accounted in the block cache
+usage, they are allocated anyway and this can lead to too much memory
+usage. On the other hand, if they are accounted in the block cache
+usage, then the block cache has less capacity for its core operations,
+the caching of data blocks.
+
+The following options control accounting for RocksDB RAM usage:
 
 ```
---rocksdb.cache-index-and-filter-blocks`
+--rocksdb.cache-index-and-filter-blocks
 --rocksdb.cache-index-and-filter-blocks-with-high-priority
+--rocksdb.reserve-file-metadata-memory (>= 3.11)
+--rocksdb.reserve-table-builder-memory
+--rocksdb.reserve-table-reader-memory
+```
+
+They are for Bloom filter and block indexes, file metadata, table
+building (RAM usage for building SST files, this happens when flushing
+memtables to level 0 SST files, during compaction and on recovery)
+and table reading (RAM usage for read operations) respectively.
+
+There are additional options you can enable to avoid that the index and filter
+blocks get evicted from cache:
+
+```
 --rocksdb.pin-l0-filter-and-index-blocks-in-cache
 --rocksdb.pin-top-level-index-and-filter
 ```
 
-Also see:
+The first does level 0 Bloom filters and index blocks and its default is
+`false`. The second does top level of partitioned index blocks and Bloom
+filters into the cache, its default is `true`.
+
+The block cache basically trades increased RAM usage for less disk I/O, so its
+size does not only affect memory usage, but can also affect read performance.
+
+See also:
 - [RocksDB Server Options](programs-arangod-options.html#rocksdb)
 - [Write Buffer Manager](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager){:target="_blank"}
 
+
+Transactions
+------------
+
+Before commit, RocksDB builds up transaction data in RAM. This happens
+in so-called "memtables" and "write batches". Note that from Version
+3.12 on this memory usage is accounted for in writing AQL queries and
+other write operations, and can thus be limited there. When there are
+many or large open transactions, this can sum up to a large amount of
+RAM usage.
+
+Once a transaction is committed and data is supposed to written out to
+level 0 SST files, there is RAM usage for table builders.
+
+As mentioned above in the block cache section, RAM usage for table
+builders can be accounted for with the block cache, if
+`--rocksdb.reserve-table-builder-memory` is set to `true`.
```

Comment:
I don't think this should be mentioned here like this.
Transaction are built up entirely in memory before they are committed, and then the data from memory is written to the WAL and memtables. The memtables will only be flushed if their size or number exceeds whatever configured value there is. And only upon memtable flushing we will need a table builder.
If a transaction does not commit or its commit does not trigger a flush, there won't be any table builders involved.
Table builders will be used during flushes and compaction though.

Reply:
Removed. The option is mentioned in the block cache section, so we should be good.